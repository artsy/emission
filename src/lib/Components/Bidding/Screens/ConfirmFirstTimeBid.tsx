import React from "react"
import { NativeModules, TouchableWithoutFeedback, View } from "react-native"
const Emission = NativeModules.Emission || {}
import { createFragmentContainer, graphql } from "react-relay"
import styled from "styled-components/native"

import { Schema, screenTrack, track } from "../../../utils/track"

import { Flex } from "../Elements/Flex"
import { Serif14, SerifItalic14, SerifSemibold14, SerifSemibold18 } from "../Elements/Typography"

import { BiddingThemeProvider } from "../Components/BiddingThemeProvider"
import { BidInfoRow } from "../Components/BidInfoRow"
import { Button } from "../Components/Button"
import { Checkbox } from "../Components/Checkbox"
import { Container } from "../Components/Containers"
import { Divider } from "../Components/Divider"
import { Timer } from "../Components/Timer"
import { Title } from "../Components/Title"

import SwitchBoard from "lib/NativeModules/SwitchBoard"

import { BillingAddress } from "./BillingAddress"
import { ConfirmBidProps } from "./ConfirmBid"

import { Colors } from "lib/data/colors"
import stripe from "tipsi-stripe"
import { CreditCardForm } from "./CreditCardForm"

stripe.setOptions({
  publishableKey: Emission.stripePublishableKey,
  // merchantId: "MERCHANT_ID", // Optional
  // androidPayMode: "test", // Android only
})

// values from the Tipsi PaymentCardTextField component
export interface PaymentCardTextFieldParams {
  number: string
  expMonth: string
  expYear: string
  cvc: string
}

export interface Address {
  fullName: string
  addressLine1: string
  addressLine2?: string
  city: string
  state: string
  postalCode: string
}

interface ConfirmBidState {
  billingAddress?: Address
  creditCardFormParams?: PaymentCardTextFieldParams
  creditCardToken?: StripeToken // TODO: change this interface accrodingly when adapting stripe
  conditionsOfSaleChecked: boolean
  isLoading: boolean
}

interface StripeToken {
  tokenId: string
  created: number
  livemode: 1 | 0
  card: any
  bankAccount: any
  extra: any
}

@screenTrack({
  context_screen: Schema.PageNames.BidFlowConfirmBidPage,
  context_screen_owner_type: null,
})
export class ConfirmFirstTimeBid extends React.Component<ConfirmBidProps, ConfirmBidState> {
  state = {
    billingAddress: undefined,
    creditCardToken: null,
    conditionsOfSaleChecked: false,
    isLoading: false,
  }

  onPressConditionsOfSale = () => {
    SwitchBoard.presentModalViewController(this, "/conditions-of-sale?present_modally=true")
  }

  showCreditCardForm() {
    this.props.navigator.push({
      component: CreditCardForm,
      title: "",
      passProps: {
        onSubmit: this.onCreditCardAdded,
        navigator: this.props.navigator,
      },
    })
  }

  showBillingAddressForm() {
    this.props.navigator.push({
      component: BillingAddress,
      title: "",
      passProps: {
        onSubmit: this.onBillingAddressAdded,
        billingAddress: this.state.billingAddress,
        navigator: this.props.navigator,
      },
    })
  }

  onCreditCardAdded = async (params: PaymentCardTextFieldParams) => {
    // Get token here or just set state? Getting token gives us card type for free but we may not have billing addr yet (for stronger token??? Assumptions)
    const token = await stripe.createTokenWithCard(params)
    this.setState({ creditCardToken: token, creditCardFormParams: params })
  }

  onBillingAddressAdded = (values: Address) => {
    this.setState({ billingAddress: values })
  }

  @track({
    action_type: Schema.ActionTypes.Tap,
    action_name: Schema.ActionNames.BidFlowPlaceBid,
  })
  placeBid() {
    return null
  }

  maxBidPressed() {
    this.props.navigator.pop()
  }

  render() {
    const { billingAddress, creditCardToken: token } = this.state

    return (
      <BiddingThemeProvider>
        <Container m={0}>
          <Flex alignItems="center">
            <Title mb={3}>Confirm your bid</Title>
            <Timer timeLeftInMilliseconds={1000 * 60 * 20} />
          </Flex>

          <View>
            <Flex m={4} mt={0} alignItems="center">
              <SerifSemibold18>{this.props.sale_artwork.artwork.artist_names}</SerifSemibold18>
              <SerifSemibold14>Lot {this.props.sale_artwork.lot_label}</SerifSemibold14>

              <SerifItalic14 color="black60">
                {this.props.sale_artwork.artwork.title}, <Serif14>{this.props.sale_artwork.artwork.date}</Serif14>
              </SerifItalic14>
            </Flex>

            <Divider mb={2} />

            <BidInfoRow label="Max bid" value={this.props.bid.display} onPress={() => this.maxBidPressed()} />

            <Divider mb={2} />

            <BidInfoRow
              label="Credit Card"
              value={token && this.formatCard(token)}
              onPress={() => this.showCreditCardForm()}
            />

            <Divider mb={2} />

            <BidInfoRow
              label="Billing address"
              value={billingAddress && this.formatAddress(billingAddress)}
              onPress={() => this.showBillingAddressForm()}
            />

            <Divider />
          </View>

          <View>
            <Checkbox justifyContent="center">
              <Serif14 mt={2} color="black60">
                You agree to <LinkText onPress={this.onPressConditionsOfSale}>Conditions of Sale</LinkText>.
              </Serif14>
            </Checkbox>

            <Flex m={4}>
              <Button text="Place Bid" onPress={() => this.placeBid()} />
            </Flex>
          </View>
        </Container>
      </BiddingThemeProvider>
    )
  }

  private formatCard(token: StripeToken) {
    return `${token.card.brand} •••• ${token.card.last4}`
  }

  private formatAddress(address: Address) {
    return [address.addressLine1, address.addressLine2, address.city, address.state].filter(el => el).join(" ")
  }
}

const LinkText = styled.Text`
  text-decoration-line: underline;
`

export const ConfirmFirstTimeBidScreen = createFragmentContainer(
  ConfirmFirstTimeBid,
  graphql`
    fragment ConfirmFirstTimeBid_sale_artwork on SaleArtwork {
      sale {
        id
      }
      artwork {
        id
        title
        date
        artist_names
      }
      lot_label
      ...BidResult_sale_artwork
    }
  `
)
