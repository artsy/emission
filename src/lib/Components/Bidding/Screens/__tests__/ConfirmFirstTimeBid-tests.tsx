import React from "react"
import "react-native"
import * as renderer from "react-test-renderer"

import { ConfirmFirstTimeBid } from "../ConfirmFirstTimeBid"

it("renders properly", () => {
  const saleArtwork = {
    artwork: {
      title: "Meteor Shower",
      date: "2015",
      artist_names: "Makiko Kudo",
    },
    lot_label: "538",
  }
  const bg = renderer
    .create(<ConfirmFirstTimeBid sale_artwork={saleArtwork} bid={{ cents: 450000, display: "$45,000" }} />)
    .toJSON()
  expect(bg).toMatchSnapshot()
})
