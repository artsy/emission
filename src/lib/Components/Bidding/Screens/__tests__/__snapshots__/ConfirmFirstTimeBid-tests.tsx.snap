// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<View
  flex={1}
  flexDirection="column"
  justifyContent="space-between"
  m={0}
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
      },
      undefined,
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    fontSize={5}
    m={4}
    style={
      Array [
        Object {
          "fontFamily": "AGaramondPro-Semibold",
          "fontSize": 22,
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 20,
          "textAlign": "center",
        },
        undefined,
      ]
    }
    textAlign="center"
  >
    Confirm your bid
  </Text>
  <View>
    <View
      alignItems="center"
      m={4}
      style={
        Array [
          Object {
            "marginBottom": 20,
            "marginLeft": 20,
            "marginRight": 20,
            "marginTop": 20,
          },
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        fontSize={4}
        style={
          Array [
            Object {
              "fontFamily": "AGaramondPro-Semibold",
              "fontSize": 18,
            },
            undefined,
          ]
        }
      >
        Makiko Kudo
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        fontSize={2}
        style={
          Array [
            Object {
              "fontFamily": "AGaramondPro-Semibold",
              "fontSize": 14,
            },
            undefined,
          ]
        }
      >
        Lot 
        538
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        color="black60"
        ellipsizeMode="tail"
        fontSize={2}
        style={
          Array [
            Object {
              "color": "#666",
              "fontFamily": "AGaramondPro-Italic",
              "fontSize": 14,
            },
            undefined,
          ]
        }
      >
        Meteor Shower
        , 
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          fontSize={2}
          style={
            Array [
              Object {
                "fontFamily": "AGaramondPro-Regular",
                "fontSize": 14,
              },
              undefined,
            ]
          }
        >
          2015
        </Text>
      </Text>
    </View>
    <View
      border={1}
      borderBottomWidth={0}
      borderColor="black10"
      mb={2}
      style={
        Array [
          Object {
            "borderColor": "#E5E5E5",
            "borderStyle": "solid",
            "borderWidth": 1,
            "marginBottom": 5,
            "width": "100%",
          },
          undefined,
        ]
      }
      width="100%"
    />
    <View
      flexDirection="row"
      justifyContent="space-between"
      m={4}
      style={
        Array [
          Object {
            "marginBottom": 20,
            "marginLeft": 20,
            "marginRight": 20,
            "marginTop": 20,
          },
          undefined,
        ]
      }
    >
      <View
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          fontSize={3}
          style={
            Array [
              Object {
                "fontFamily": "AGaramondPro-Semibold",
                "fontSize": 16,
              },
              undefined,
            ]
          }
        >
          Max bid
        </Text>
      </View>
      <View
        alignItems="flex-end"
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          fontSize={3}
          style={
            Array [
              Object {
                "fontFamily": "AGaramondPro-Regular",
                "fontSize": 16,
              },
              undefined,
            ]
          }
        >
          $45,000
        </Text>
      </View>
    </View>
    <View
      border={1}
      borderBottomWidth={0}
      borderColor="black10"
      mb={2}
      style={
        Array [
          Object {
            "borderColor": "#E5E5E5",
            "borderStyle": "solid",
            "borderWidth": 1,
            "marginBottom": 5,
            "width": "100%",
          },
          undefined,
        ]
      }
      width="100%"
    />
    <View
      flexDirection="row"
      justifyContent="space-between"
      m={4}
      style={
        Array [
          Object {
            "marginBottom": 20,
            "marginLeft": 20,
            "marginRight": 20,
            "marginTop": 20,
          },
          undefined,
        ]
      }
    >
      <View
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          fontSize={3}
          style={
            Array [
              Object {
                "fontFamily": "AGaramondPro-Semibold",
                "fontSize": 16,
              },
              undefined,
            ]
          }
        >
          Credit card
        </Text>
      </View>
      <View
        alignItems="flex-end"
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          color="purple100"
          ellipsizeMode="tail"
          fontSize={3}
          style={
            Array [
              Object {
                "color": "#6E1EFF",
                "fontFamily": "AGaramondPro-Regular",
                "fontSize": 16,
              },
              undefined,
            ]
          }
        >
          Add
        </Text>
      </View>
    </View>
    <View
      border={1}
      borderBottomWidth={0}
      borderColor="black10"
      mb={2}
      style={
        Array [
          Object {
            "borderColor": "#E5E5E5",
            "borderStyle": "solid",
            "borderWidth": 1,
            "marginBottom": 5,
            "width": "100%",
          },
          undefined,
        ]
      }
      width="100%"
    />
    <View
      flexDirection="row"
      justifyContent="space-between"
      m={4}
      style={
        Array [
          Object {
            "marginBottom": 20,
            "marginLeft": 20,
            "marginRight": 20,
            "marginTop": 20,
          },
          undefined,
        ]
      }
    >
      <View
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          fontSize={3}
          style={
            Array [
              Object {
                "fontFamily": "AGaramondPro-Semibold",
                "fontSize": 16,
              },
              undefined,
            ]
          }
        >
          Billing address
        </Text>
      </View>
      <View
        alignItems="flex-end"
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          color="purple100"
          ellipsizeMode="tail"
          fontSize={3}
          style={
            Array [
              Object {
                "color": "#6E1EFF",
                "fontFamily": "AGaramondPro-Regular",
                "fontSize": 16,
              },
              undefined,
            ]
          }
        >
          Add
        </Text>
      </View>
    </View>
    <View
      border={1}
      borderBottomWidth={0}
      borderColor="black10"
      style={
        Array [
          Object {
            "borderColor": "#E5E5E5",
            "borderStyle": "solid",
            "borderWidth": 1,
            "width": "100%",
          },
          undefined,
        ]
      }
      width="100%"
    />
  </View>
  <View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      alignSelf="center"
      flexDirection="column"
      hitSlop={undefined}
      m={3}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "marginBottom": 10,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 10,
          },
          undefined,
        ]
      }
      testID={undefined}
    >
      <View
        alignItems="center"
        flexDirection="row"
        mb={2}
        style={
          Array [
            Object {
              "marginBottom": 5,
            },
            undefined,
          ]
        }
      >
        <View
          alignItems="center"
          bg={null}
          border="2px solid"
          borderColor="black10"
          height={20}
          justifyContent="center"
          mr={3}
          style={
            Array [
              Object {
                "borderColor": "#E5E5E5",
                "borderStyle": "solid",
                "borderWidth": 2,
                "height": 20,
                "marginRight": 10,
                "width": 20,
              },
              undefined,
            ]
          }
          width={20}
        >
          <View
            collapsable={undefined}
            duration={250}
            show={false}
            style={
              Object {
                "opacity": 0,
                "transform": Array [
                  Object {
                    "scale": 1.1,
                  },
                ],
              }
            }
          />
        </View>
        <View
          style={
            Array [
              Object {},
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            color="black60"
            ellipsizeMode="tail"
            fontSize={3}
            mt={2}
            style={
              Array [
                Object {
                  "color": "#666",
                  "fontFamily": "AGaramondPro-Regular",
                  "fontSize": 16,
                  "marginTop": 5,
                  "textAlign": "center",
                },
                undefined,
              ]
            }
            textAlign="center"
          >
            Agree to 
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              onPress={[Function]}
              style={
                Array [
                  Object {
                    "textDecorationLine": "underline",
                  },
                  undefined,
                ]
              }
            >
              Conditions of Sale
            </Text>
            .
          </Text>
        </View>
      </View>
    </View>
    <View
      m={4}
      style={
        Array [
          Object {
            "marginBottom": 20,
            "marginLeft": 20,
            "marginRight": 20,
            "marginTop": 20,
          },
          undefined,
        ]
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 1)",
            "borderColor": "rgba(0, 0, 0, 1)",
            "borderWidth": 1,
            "height": 46,
            "justifyContent": "center",
          }
        }
        testID={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontSize": 12,
              },
              Object {
                "color": "rgba(255, 255, 255, 1)",
                "fontFamily": "AGaramondPro-Regular",
                "fontSize": 14,
                "opacity": 1,
              },
              Object {
                "fontFamily": "AvantGardeGothicITC",
              },
            ]
          }
        >
          PLACE BID
        </Text>
      </View>
    </View>
  </View>
</View>
`;
